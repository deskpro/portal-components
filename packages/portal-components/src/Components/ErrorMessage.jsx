import React from 'react';
import PropTypes from 'prop-types';
import { Field, getIn } from 'formik';

const ErrorMessage = ({ name }) => (
  <Field
    name={name}
    render={({ form }) => {
      const error = getIn(form.errors, name);
      const touch = getIn(form.touched, name);
      return touch && error ? error : null;
    }}
  />
);

ErrorMessage.propTypes = {
  name: PropTypes.string.isRequired
};

export default ErrorMessage;
