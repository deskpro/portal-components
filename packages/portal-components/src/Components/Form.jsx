import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => <FormikForm {...props} />;

export default Form;
