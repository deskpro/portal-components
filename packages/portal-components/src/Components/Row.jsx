import React from 'react';
import PropTypes from 'prop-types';

const Row = ({
  children,
  ...props
}) => {
  return <div {...props}>{children}</div>
};

Row.propTypes = {
  /**
   * Children to render.
   */
  children:  PropTypes.node
};

export default Row;