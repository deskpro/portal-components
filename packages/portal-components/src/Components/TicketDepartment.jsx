import React from 'react';
import PropTypes from 'prop-types';
import DropDown from './Choices/DropDown';

class TicketDepartment extends React.Component {
  static propTypes = {
    departments:  PropTypes.array.isRequired,
    handleChange: PropTypes.func.isRequired
  };

  render() {
    const { departments, handleChange } = this.props;
    return (
      <DropDown
        name="department"
        label="Department"
        options={departments}
        handleChange={handleChange}
        clearable={false}
      />
    );
  }
}

export default TicketDepartment;
