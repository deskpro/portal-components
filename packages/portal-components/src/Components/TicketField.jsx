import React from 'react';
import PropTypes from 'prop-types';
import FileUpload from './Inputs/FileUpload';
import Text from './Inputs/Text';
import Textarea from './Inputs/Textarea';

class TicketField extends React.Component {
  static propTypes = {
    field: PropTypes.object.isRequired,
  };

  renderCustomField = () => 'Custom field';

  render() {
    const { field } = this.props;
    if (field.get('field_id').match(/^ticket_field/)) {
      return this.renderCustomField();
    }
    switch (field.get('field_id')) {
      case 'message':
        return <Textarea name="message" label="Message" />;
      case 'attachments':
        return <FileUpload name="attachments" label="Attachments" />;
      default:
        return (
          <Text
            name={field.get('field_id')}
            label={field.get('field_id')}
          />
        );
    }
  }
}

export default TicketField;
