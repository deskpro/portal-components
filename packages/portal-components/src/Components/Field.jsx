import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Field as FormikField, getIn } from 'formik';
import newid from '@deskpro/js-utils/dist/newid';
import ErrorMessage from './ErrorMessage';

class Field extends React.Component {
  static propTypes = {
    name:        PropTypes.string.isRequired,
    label:       PropTypes.string,
    description: PropTypes.string,
    id:          PropTypes.string,
    children:    PropTypes.node
  };

  static defaultProps = {
    label:       '',
    description: '',
    id:          null,
    children:    null,
  };

  constructor(props) {
    super(props);

    this.id = props.id;
    if (!this.id) {
      this.id = newid('field_');
    }
  }

  renderField = () => {
    const {
      name,
      children,
      ...props
    } = this.props;

    return (
      <FormikField
        id={this.id}
        name={name}
        type={this.type}
        className="dp-pc_input"
        {...props}
      >
        {children}
      </FormikField>
    );
  };

  renderDescription = () => {
    const {
      description
    } = this.props;

    return (
      <span className="dp-pc_description">{description}</span>
    );
  };

  renderLabel = () => {
    const {
      label,
    } = this.props;
    const htmlFor = this.id;
    /* eslint-disable jsx-a11y/label-has-for */
    return (
      <label className="dp-pc_label" htmlFor={htmlFor}>
        {label}
      </label>
    );
    /* eslint-enable jsx-a11y/label-has-for */
  };

  render() {
    const {
      name,
    } = this.props;
    return (
      <FormikField
        name={name}
        render={({ form }) => {
          const error = getIn(form.errors, name);
          const touch = getIn(form.touched, name);
          const value = getIn(form.values, name);
          return (
            <div className={classNames('dp-pc_field', { 'dp-pc_error': touch && error, 'dp-pc_empty': !value })}>
              {this.renderField(form)}
              <span className="indicator" />
              { touch && error ? <ErrorMessage name={name} form={form} /> : this.renderDescription() }
              {this.renderLabel()}
            </div>
          );
        }}
      />
    );
  }
}

export default Field;
