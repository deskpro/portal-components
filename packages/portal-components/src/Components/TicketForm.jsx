import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import Yup from 'yup';
import Form from './Form';
import Submit from './Submit';
import TicketField from './TicketField';
import TicketDepartment from './TicketDepartment';

class TicketForm extends React.Component {
  static propTypes = {
    deskproLayout: PropTypes.object.isRequired,
    departments:   PropTypes.object.isRequired,
    onSubmit:      PropTypes.func.isRequired,
    department:    PropTypes.number,
    fileUploadUrl: PropTypes.string.isRequired,
    csrfToken:     PropTypes.string.isRequired,
    showHover:     PropTypes.bool,
  };

  static defaultProps = {
    department: null,
    showHover:  true,
  };

  constructor(props) {
    super(props);
    this.state = {
      department: props.department
    };
  }

  getLayout = () => {
    const layout = this.props.deskproLayout.find(d => d.get('department') === this.state.department);
    if (layout) {
      return layout;
    }
    return this.props.deskproLayout.find(d => d.get('department') === null);
  };

  getInitialValues = () => {
    const initialValues = {};
    this.getLayout().get('fields', []).forEach((field) => {
      if (field.get('field_id') === 'department') {
        initialValues.department = this.state.department;
      } else {
        initialValues[field.get('field_id')] = field.getIn(['data', 'default_value'], '');
      }
    });
    return initialValues;
  };

  getValidationSchema = () => {
    // return lazy(obj)
    const shape = {};
    this.getLayout().get('fields', []).forEach((field) => {
      let validationRule = Yup.string();
      if (field.get('required', false) || field.getIn(['data', 'options', 'validation_type']) === 'required') {
        validationRule = validationRule.required('Field is required');
      }
      const regex = field.getIn(['data', 'options', 'regex']);
      if (regex) {
        validationRule = validationRule.matches(new RegExp(regex.split('/')[1]), {
          message:            'Field is invalid',
          excludeEmptyString: field.getIn(['data', 'options', 'regex_required'])
        });
      }
      shape[field.get('field_id')] = validationRule;
    });
    return Yup.object().shape(shape);
  };

  handleDepartmentChange = (department) => {
    this.setState({
      department
    });
  };

  renderFields = () => {
    const { departments, fileUploadUrl, csrfToken } = this.props;
    return this.getLayout().get('fields', []).map((field) => {
      switch (field.get('field_id')) {
        case 'department':
          return (
            <TicketDepartment
              key="department"
              departments={departments}
              handleChange={this.handleDepartmentChange}
            />
          );
        default:
          return (
            <TicketField
              key={field.get('field_id')}
              field={field}
              fileUploadUrl={fileUploadUrl}
              csrfToken={csrfToken}
            />
          );
      }
    });
  };

  render() {
    const { showHover } = this.props;
    return (
      <Formik
        initialValues={this.getInitialValues()}
        onSubmit={this.props.onSubmit}
        validationSchema={this.getValidationSchema()}
      >
        {() => (
          <Form
            showHover={showHover}
          >
            {this.renderFields()}
            <Submit>Submit</Submit>
          </Form>
        )}
      </Formik>
    );
  }
}

export default TicketForm;
