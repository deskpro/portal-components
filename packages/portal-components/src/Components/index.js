import Checkboxes from './Choices/Checkboxes';
import DropDown from './Choices/DropDown';
import MultipleDropDown from './Choices/MultipleDropDown';
import Radio from './Choices/Radio';

import Progress from './Elements/Progress';

import Checkbox from './Inputs/Checkbox';
import DatePicker from './Inputs/DatePicker';
import DateTimePicker from './Inputs/DateTimePicker';
import Email from './Inputs/Email';
import FileUpload from './Inputs/FileUpload';
import LikertScale from './Inputs/LikertScale';
import Number from './Inputs/Number';
import Password from './Inputs/Password';
import Search from './Inputs/Search';
import Tel from './Inputs/Tel';
import Text from './Inputs/Text';
import Textarea from './Inputs/Textarea';
import Url from './Inputs/Url';

import Col from './Col';
import Form from './Form';
import Row from './Row';
import Submit from './Submit';
import TicketForm from './TicketForm';

export {
  Checkboxes,
  DropDown,
  MultipleDropDown,
  Radio,

  Progress,

  Checkbox,
  DatePicker,
  DateTimePicker,
  Email,
  FileUpload,
  LikertScale,
  Number,
  Password,
  Search,
  Tel,
  Text,
  Textarea,
  Url,

  Col,
  Form,
  Row,
  Submit,
  TicketForm
};

