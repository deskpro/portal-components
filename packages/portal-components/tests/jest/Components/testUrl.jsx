import React from 'react';
import renderer from 'react-test-renderer';
import { withFormik } from 'formik';
import { Form, Url } from 'Components';

describe('>>> Url --- Snapshot', () => {
  it('+++capturing Snapshot of Url', () => {
    const App = () => (
      <Form>
        <Url name="url" />
      </Form>
    );
    const FormikApp = withFormik({
      mapPropsToValues({
        url
      }) {
        return {
          url: url || 'www.deskpro.com'
        };
      },
      handleSubmit(values, { setSubmitting }) {
        setSubmitting(false);
      }
    })(App);
    const renderedValue = renderer.create(<FormikApp />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
