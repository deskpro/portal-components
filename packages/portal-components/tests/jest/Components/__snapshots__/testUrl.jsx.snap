// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Url --- Snapshot +++capturing Snapshot of Url 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="dp-pc_field"
  >
    <input
      errors={Object {}}
      name="url"
      onBlur={[Function]}
      onChange={[Function]}
      touched={Object {}}
      type="url"
      value="www.deskpro.com"
    />
  </div>
</form>
`;
